### PR TITLE
Prevent accidental treatment of GroupReference as real PlanNode

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

**Comment by findepi** [comment]
cc @martint

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 1]
```diff
@@ -219,5 +220,11 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<PlanNode> context)
 
             return new ProjectNode(idAllocator.getNextId(), rewrittenNode.getSource(), assignments);
         }
+
+        @Override
+        public PlanNode visitGroupReference(GroupReference node, RewriteContext<PlanNode> context)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java]: It is strange that we don't resolve `GroupReferences` in `ExtendProjectionRewriter` while other visitors do resolve `GroupReferences`. It's feels fishy. Is that correct behaviour @kokosing?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchingVisitor.java [hunk 2]
```diff
@@ -106,7 +107,7 @@ protected MatchResult visitPlan(PlanNode node, PlanMatchPattern pattern)
         }
 
         // Leaf node in the plan.
-        if (node.getSources().isEmpty()) {
+        if (sources(node).isEmpty()) {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchingVisitor.java]: undo... lets change `visitGroupReference` method

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java [hunk 3]
```diff
@@ -463,7 +464,7 @@ public PlanMatchPattern(List<PlanMatchPattern> sourcePatterns)
                 states.add(new PlanMatchingState(ImmutableList.of(this)));
             }
         }
-        if (node.getSources().size() == sourcePatterns.size() && matchers.stream().allMatch(it -> it.shapeMatches(node))) {
+        if (sources(node).size() == sourcePatterns.size() && matchers.stream().allMatch(it -> it.shapeMatches(node))) {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: I would prefer to handle `GroupReferenceNode` in `if` here

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java]: It seems that all other visitors are not called from rules, are they?

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java]: @sopel39 i retained the existing behavior. I didn't understand that though

**Comment by findepi** [comment]
AC

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchingVisitor.java [hunk 4]
```diff
@@ -92,7 +93,7 @@ public MatchResult visitGroupReference(GroupReference node, PlanMatchPattern pat
         if (match.isMatch()) {
             return match;
         }
-        return visitPlan(node, pattern);
+        return matchLeaf(node, pattern, pattern.shapeMatches(node));
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchingVisitor.java]: what is leaf in this context?

Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchingVisitor.java]: This means: if `resolve(group)` did not match the expected pattern, let's try to match `group` itself. The un-resolved group can match only as a leaf.

**Comment by findepi** [comment]
Added one fixup. Otherwise a test introduced in https://github.com/prestodb/presto/pull/9993 would fail after this is merged.

**Pull request merged**